src/main/java/com/coreos/jetcd/Lease.java

```diff
@@ -50,57 +33,37 @@
   CompletableFuture<LeaseKeepAliveResponse> keepAliveOnce(long leaseId);
 
   /**
-   * set LeaseHandler for lease.
+   * keep the given lease alive forever.
    *
-   * @param leaseId id of the lease to set handler
-   * @param leaseHandler the handler for the lease
-   * @throws NoSuchLeaseException if lease do not exist
+   * @param leaseId lease to be keep alive forever.
+   * @return a KeepAliveListener that listens for KeepAlive responses.
    */
-  void setLeaseHandler(long leaseId, LeaseHandler leaseHandler)
-      throws NoSuchLeaseException;
+  KeepAliveListener keepAlive(long leaseId);
 
   /**
-   * Init the request stream to etcd.
-   * start schedule to keep heartbeat to keep alive and remove dead leases.
-   *
-   * @throws IllegalStateException if the service is running already
+   * close Lease client and release its resources.
    */
-  void startKeepAliveService() throws IllegalStateException;
+  void close();
 
   /**
-   * end the schedule for keep alive and remove dead leases.
-   *
-   * @throws IllegalStateException if the service is not running yet
+   * KeepAliveListener listens for LeaseKeepAliveResponse of a given leaseID.
    */
-  void closeKeepAliveService() throws IllegalStateException;
-
-  /**
-   * It hints the state of the keep alive service.
-   *
-   * @return whether the keep alive service is running.
-   */
-  boolean isKeepAliveServiceRunning();
-
-  /**
-   * This interface is called by Etcd KeepAlive client to notify user about lease expiration and
-   * exception.
-   */
-  interface LeaseHandler {
+  interface KeepAliveListener {
```

Comment:
For me, Watcher acts like a active agent. It watches stuff that changes. Listener is more a passive agent where it just receives whatever other tells it to do. In that sense, I am not sure if Listener style make sense for Watch; even though under the hood both just receives events from etcd server.